pkg/build/api/types.go

```diff
@@ -76,8 +76,72 @@ type BuildSpec struct {
 	// scheduled in the system, that the build may be active on a node before the
 	// system actively tries to terminate the build; value must be positive integer
 	CompletionDeadlineSeconds *int64
+
+	// Info is a placeholder that can store multiple reasons why a build was
+	// triggered. It should cover for a future implementation of level triggered
+	// builds where you can multiple reasons why a build was triggered.
+	Info []*BuildTriggerInfo
+}
+
+// BuildTriggerInfo contains information about why/who/what triggered a build.
+type BuildTriggerInfo struct {
+	// Details is a placeholder for  information about who triggered a build if
+	// it's either a user, a token or something else
+	Details BuildTriggerDetails
```

Comment:
except for generic and github triggers, it looks exactly like you described it.
i didn't update this PR yet because i changed some stuff and it was broken, but will do asap.
what i did add, was for configchange i now keep the lastversion value of the bc, which i'm not that sure that adds a lot of value.
for imagechange i added the from ref like you suggested the first time.
as far as getting user credentials, i'm not sure how to do that. i traced some functions to kubernetes but i might need some help from @rhcarvalho to use them properly
still not sure how to get user details or information for whenever the user clicks the rebuild button in the webconsole.